File: 알고리즘/삽입정렬/insertionSort.py
#  방법1 - while + shift
def insertion_sort(arr):
    for i in range(1, len(arr)):
        key = arr[i]
        j = i - 1
        while j >= 0 and arr[j] > key:
            arr[j + 1] = arr[j]
            j -= 1
        arr[j + 1] = key
    return arr

#  방법2 - for + swap
def insertion_sort(arr):
    for i in range(1, len(arr)):
        for j in range(i,0,-1):
            if arr[j] < arr[j-1]:
                arr[j],arr[j-1] = arr[j-1],arr[j]
            else:
                break
    return arr
   

# 예시
data = [5, 2, 4, 6, 1, 3]
print("정렬 결과:", insertion_sort(data))

File: 알고리즘/삽입정렬/InsertionSort.java
package 알고리즘.삽입정렬;

public class InsertionSort {

    // 방법1 - while + shift 
    public static void insertionSortShift(int[] arr) {
        for (int i=1; i< arr.length; i++){
            int key = arr[i];
            int j = i-1;
            while (j >=0 && arr[j] > key){
                arr[j+1] = arr[j]; // shift
                j--;
            }
            arr[j+1] = key; // insert
        }
    }

    // 방법2 - for + swap 
    public static void insertionSortSwap(int[] arr) {
        for (int i=1; i<arr.length; i++){
            for (int j=i-1; j>=0; j--){
                if(arr[j] < arr[j-1]){
                    int temp = arr[j];
                    arr[j] = arr[j-1];
                    arr[j-1] = temp;
                }else{
                    break; // 이 앞은 이미 정렬된 상태
                }
            }
        }
    }

    public static void main(String[] args){
        int[] arr = {5,1,4,2,8};
        insertionSortShift(arr);
        // insertionSortSwap(arr);
        System.out.print("정렬 후 배열 : ");
        for(int num : arr){
            System.out.print(num + " ");
        }
    }
}
